Verify maker & taker fee transactions via Mempool lookup

jmacxx opened the pull request:
Closes #5119

- Queries mempool at checkOfferAvailability stage (take offer).
- Validates maker transaction and trading fee.
- Indicates status at UI level (see screenshots below)
- Queries mempool at trade initiation (step 1&2) before fiat payment initiated.
- Maker validates taker transaction and trading fee.
- Informs user to open a dispute if the trading fees were not paid
- Arbitrator ticket informs about the confirmation status of delayed payout

---

**Screenshots**

https://imgur.com/a/tSaUgRL

---

**Safeguards**

Can be disabled universally via a filter, and individually via config setting.  

In regtest mode the API calls are bypassed, returning success status.

---


**Accuracy**

The fee schedule is set via DAO parameters.  Therefore checking that the correct fee was paid involves comparing the block height of an offer Tx to the DAO parameters which were relevant at that time.

I observed some variance in fee payments, perhaps because different users nodes may have been out of consensus with the DAO state.  

The validation code has a threshold at which it considers fees paid to be acceptable.  Currently this is set at 95%, i.e. if the fee paid is >= 95% of what was expected, the validation passes.

Very old offers (of which there are a handful still out there), use fee addresses which are not known but perhaps were used prior to the DAO.  So a threshold of blockheight 599999 is used such that the unknown fee addresses are allowed to pass when the offer Tx is older than the threshold.  Alternatively we could add [1PUXU1M...](https://mempool.space/address/1PUXU1MQ82JC3Hx1NN5tZs3BaTAJVg72MC) and [18GzH11...](https://mempool.space/address/18GzH11T5h2fpvUoBJDub7MgNJVw3FfqQ8) to the validator's list of accepted fee addresses.

    private final double feeTolerance = 0.95;     // we expect fees to be at least 95% of target
    private final long blockTolerance = 599999L;  // allow really old offers with weird fee addresses

Out of ~450 current offers ~9 are flagged as having underpaid the trading fee.


----

**Checks own open offers**

To implement the requirement that one's own offers be checked I integrated a validation in the TriggerPriceService so that if an offer is found to be bad it gets automatically disabled.  This can help prevent the issues we've been seeing of [offers that do not get published to the mempool for hours/days sometimes](https://github.com/bisq-network/bisq/issues/5151).  Now if an offer Tx is not found, it will get disabled and the user will be motivated to investigate.

----

**Testing**

Tests were written in `TxValidatorTest` to check the various error scenarios, be it underpaid BSQ, BTC, unknown fee address, or unknown tx.  Most of the tests use data from real-world scenarios like the dust-fee scam.

----

### PR Checklist:

- [X] Validate offer maker tx and fees.
- [X] Arbitrator checks that delayed payout tx is confirmed.
- [x] Verify deposit tx inputs match maker/taker outputs.
- [x] Maker verify taker fee payment.
- [x] Validates own open offers, disables ones that are bad.

Conza88 commented:
I thought it did this already 😳

ripcurlx commented:
> I thought it did this already 😳

Only for transactions that are linked to your wallet (e.g. Bisq fee tx of the trading peer wasn't checked)

Conza88 commented:
> * If there is a problem it warns user to back out (user can still continue).

Would there be a scenario when that was ever advisable? Or worth doing?

Commit:
Validate maker/taker fees using mempool lookup
apply @chimp1984 patch.txt code review suggestions
taker tx check moved to trade step 2, after confirmation
solve issue with calculating expected fees for unconfirmed tx
resolve conflict with PR5207 (Disputes UI)
check new offers after 1 block; check Json string not null; warn -> info
remove unused parameter
remove debugging log.warn message

chimp1984 commented on core/src/main/java/bisq/core/provider/mempool/TxValidator.java:
```diff
@@ -0,0 +1,379 @@
+        long feeValue = jsonVIn0Value.getAsLong() - jsonFeeValue.getAsLong();
+        // if the first output (BSQ) is greater than the first input (BSQ) include the second input (presumably BSQ)
+        if (jsonFeeValue.getAsLong() > jsonVIn0Value.getAsLong()) {
+            // in this case 2 or more UTXOs were spent to pay the fee:
```
The code handles only the case of 1 or 2 BSQ inputs. Maybe also add a comment that we building on the concention that BSQ inputs are the before BTC inputs, which is not a consensus rule (for outputs it is) but our code handles it like that, so for the fee check that is good enough as we do not expect that ppl use custom txs for those txs.
It could be also a rare case that there is no BSQ change output. That can be if the input is exactly the amount to be burned or if the change is < dust, then it gets added to miner fee.

Another issue is that we do not check if the burned satoshis are BSQ or BTC. That would require that the tx is confirmed as only at confirmation the BSQ explorers recognize a BSQ tx. As confirmation will happen later we need to do that check after the deposit tx is confirmed in the trade protocol. So before the buyer sends the fiat he will check the takers BSQ fee it the BSQ explorer returns it as a valid BSQ tx and then before the seller confirms receipt to check the makers fee tx.

chimp1984 commented on core/src/main/java/bisq/core/provider/mempool/TxValidator.java:
```diff
@@ -0,0 +1,379 @@
+        JsonObject jsonVin0 = jsonVin.get(0).getAsJsonObject();
+        JsonObject jsonVout0 = jsonVout.get(0).getAsJsonObject();
+        JsonElement jsonVIn0Value = jsonVin0.getAsJsonObject("prevout").get("value");
+        JsonElement jsonFeeValue = jsonVout0.get("value");
```
The BSQ fee check is tricky. Maybe a better approach is to postpone it directly to the trade protocol states when the deposit tx is confirmed and use the BSQ explorer for delivering the fee.
In BsqWalletService.completeBsqTradingFeeTx there have been added a while ago a change which made the input 0 a BTC output in case there is no BSQ output. I think that change is not needed (at least I tested it without and it worked, but don't want to touch that at the moment...). So to take that additional option into account all gets more complicated....

chimp1984 commented on core/src/main/java/bisq/core/trade/TradeDataValidation.java:
```diff
@@ -355,6 +355,27 @@ public static void validatePayoutTxInput(Transaction depositTx,
     }
 
+    public static void validateDepositInputs(Trade trade) throws InvalidTxException {
+        // assumption: deposit tx always has 2 inputs, the maker and taker
```
Correct.

chimp1984 commented on core/src/main/java/bisq/core/trade/TradeDataValidation.java:
```diff
@@ -355,6 +355,27 @@ public static void validatePayoutTxInput(Transaction depositTx,
+        String contractTakerTxId = trade.getContract().getTakerFeeTxID();
+        boolean makerFirstMatch = contractMakerTxId.equalsIgnoreCase(txIdInput0) && contractTakerTxId.equalsIgnoreCase(txIdInput1);
+        boolean takerFirstMatch = contractMakerTxId.equalsIgnoreCase(txIdInput1) && contractTakerTxId.equalsIgnoreCase(txIdInput0);
+        if (!makerFirstMatch && !takerFirstMatch) {
```
The order is determined by buyer/seller role (not maker/taker) but your check here is sufficient...

chimp1984 commented on core/src/main/java/bisq/core/provider/mempool/MempoolService.java:
```diff
@@ -0,0 +1,250 @@
+import javax.annotation.Nullable;
+
+@Slf4j
+public class MempoolService {
```
I assume it should be a @Singleton . At the moment we get created 2 instances.

chimp1984 commented on core/src/main/java/bisq/core/provider/mempool/MempoolService.java:
```diff
@@ -0,0 +1,250 @@
+        this.filterManager = filterManager;
+        this.daoFacade = daoFacade;
+        this.daoStateService = daoStateService;
+        this.btcFeeReceivers = getAllBtcFeeReceivers();
```
As we rely here on the daostate we need to fill btcFeeReceivers via a onAllServicesInitialized method.

chimp1984 commented on core/src/main/java/bisq/core/provider/mempool/MempoolService.java:
```diff
@@ -0,0 +1,250 @@
+            @Override
+            public void onSuccess(@Nullable String jsonTxt) {
+                outstandingRequests--;
+                UserThread.execute(() -> {
```
As we get called from another thread here its more safe to move the outstandingRequests update into the userthread mapped runnable as well.

chimp1984 commented on core/src/main/java/bisq/core/provider/mempool/MempoolService.java:
```diff
@@ -0,0 +1,250 @@
+                if (theRequest.switchToAnotherProvider()) {
+                    retryValidateOfferMakerTx(theRequest, txValidator, resultHandler);
+                } else {
+                    UserThread.execute(() -> {
```
I think here as well its more safe to move all code into the runnable to avoid potential threading issues.

chimp1984 commented on desktop/src/main/java/bisq/desktop/main/offer/takeoffer/TakeOfferView.java:
```diff
@@ -898,6 +899,21 @@ private void addButtons() {
+    private void nextStepCheckMakerTx() {
+        // the tx validation check has had plenty of time to complete, but if for some reason it has not returned
+        // we continue anyway since the check is not crucial.
+        // note, it would be great if there was a real tri-state boolean we could use here, instead of -1, 0, and 1
```
Using Optional<Boolean> would be an option to handle that. Another to use boxed Boolean type and check for null but Optional is more clear IMO.

chimp1984 commented on desktop/src/main/java/bisq/desktop/main/offer/takeoffer/TakeOfferViewModel.java:
```diff
@@ -462,6 +463,7 @@ private void updateButtonDisableState() {
         boolean inputDataValid = isBtcInputValid(amount.get()).isValid
                 && dataModel.isMinAmountLessOrEqualAmount()
                 && !dataModel.isAmountLargerThanOfferAmount()
+                && dataModel.mempoolStatus.get() >= 0
```
Do we want to block in case response is slow (tor can be slow)?

ghost commented:
:+1:  Thanks @chimp1984 for the review & suggestions.  I applied your changes, and will think about the suggestion of checking BSQ fees using a BSQ explorer.

ripcurlx commented:
@jmacxx I'm reviewing this PR right now and think it would be great to get it into v1.6.0. Could you please resolve the conflicts mentioned and I'm happy to merge it after a positive review.

ripcurlx commented on core/src/main/java/bisq/core/offer/TriggerPriceService.java:
```diff
@@ -152,6 +156,20 @@ private void checkPriceThreshold(MarketPrice marketPrice, OpenOffer openOffer) {
+            // check the mempool if it has not been done before
+            if (openOffer.getMempoolStatus() < 0 && mempoolService.canRequestBeMade()) {
+                mempoolService.validateOfferMakerTx(openOffer.getOffer().getOfferPayload(), (txValidator -> {
+                    openOffer.setMempoolStatus(txValidator.isFail() ? 0 : 1);
```
So if it fails to validate the tx the offer will be deactivated after the next price update (~60 sec later), right?

ripcurlx commented:
I just got following error while trying to open a dispute as mediator:
```
März-17 15:33:03.377 [JavaFX Application Thread] ERROR bisq.common.setup.CommonSetup: Uncaught Exception from thread JavaFX Application Thread 
März-17 15:33:03.377 [JavaFX Application Thread] ERROR bisq.common.setup.CommonSetup: throwableMessage= null 
März-17 15:33:03.378 [JavaFX Application Thread] ERROR bisq.common.setup.CommonSetup: throwableClass= class java.lang.NullPointerException 
März-17 15:33:03.381 [JavaFX Application Thread] ERROR bisq.common.setup.CommonSetup: Stack trace:
java.lang.NullPointerException
	at bisq.core.provider.mempool.TxValidator.initialSanityChecks(TxValidator.java:260)
	at bisq.core.provider.mempool.TxValidator.parseJsonValidateTx(TxValidator.java:131)
```
This happens as `jsonTxt` is "" on Regtest.

ripcurlx commented on desktop/src/main/java/bisq/desktop/main/offer/takeoffer/TakeOfferDataModel.java:
```diff
@@ -241,6 +254,15 @@ void initWithData(Offer offer) {
+            mempoolStatus.setValue(txValidator.isFail() ? 0 : 1);
+            if (txValidator.isFail()) {
+                mempoolStatusText = txValidator.toString();
+                log.warn("Mempool check of OfferFeePaymentTxId returned errors: [{}]", mempoolStatusText);
```
This should be 
```suggestion
                log.info("Mempool check of OfferFeePaymentTxId returned errors: [{}]", mempoolStatusText);
```

ripcurlx replied:
Recognized it as I'm always checking for warning in the log :wink:

ripcurlx commented:
It wasn't changed in this PR, but I missed it in the other one. 
Please don't use `version-new` as class to highlight this text, as there might be a problem in the future if someone changes this class. It also shows a hand icon which indicates that this underlined label is clickable.

`tradePeriodEnd.getStyleClass().add("version-new"); // highlight field when the trade period is still active`

Better would be e.g. to create a new class `.alert` without underline and hand, so it doesn't look like a link.
```
.alert {
    -fx-text-fill: -bs-rd-error-red;
}

```
![Bildschirmfoto 2021-03-17 um 16 08 06](https://user-images.githubusercontent.com/170962/111490055-fdbff200-873a-11eb-8265-3d8bd74b887c.png)

ripcurlx commented:
And one more thing I missed in the other PR and feels a little bit weird for the user. If a dispute is closed you get the notification popup. When you click on it, it opens the support section and the issue is selected. But it is greyed out and users might not know what to do. Maybe it is better to open in this case automatically the same action as a double click would trigger. WDYT?

ripcurlx commented:
Also it would be good to show a (1) badge over the chat icon for the system message, so people are looking into it. In the past the chat was open by default in-line. So at least a notification would be good.

ripcurlx commented on desktop/src/main/java/bisq/desktop/main/offer/takeoffer/TakeOfferDataModel.java:
```diff
@@ -173,6 +185,7 @@ protected void activate() {
             tradeManager.checkOfferAvailability(offer,
                     false,
                     () -> {
+                log.warn("OfferAvailability completed!");
```
I think something went wrong with commit squashing as the `log.warn` is in again.

Commit:
Merge branch 'master' into mempool_check_maker_tx

ripcurlx reviewed (approved):
ACK - tested it on Regtest and Mainnet. So far everything is looking fine besides the minor issues that are fixed now.
